src/objective-c/GRPCClient/GRPCInterceptor.m

```diff
@@ -19,146 +19,276 @@
+  }
+  id<GRPCInterceptorInterface> copiedNextInterceptor = _nextInterceptor;
+  dispatch_async(copiedNextInterceptor.dispatchQueue, ^{
+    [copiedNextInterceptor finish];
+  });
 }
 
 - (void)cancelNextInterceptor {
-  if (_nextInterceptor != nil) {
-    id<GRPCInterceptorInterface> copiedNextInterceptor = _nextInterceptor;
-    dispatch_async(copiedNextInterceptor.requestDispatchQueue, ^{
-      [copiedNextInterceptor cancel];
-    });
+  if (_nextInterceptor == nil && !_shutDown) {
+    [self createNextInterceptor];
   }
+  if (_nextInterceptor == nil) {
+    return;
+  }
+  id<GRPCInterceptorInterface> copiedNextInterceptor = _nextInterceptor;
+  dispatch_async(copiedNextInterceptor.dispatchQueue, ^{
+    [copiedNextInterceptor cancel];
+  });
 }
 
 /** Notify the next interceptor in the chain to receive more messages */
 - (void)receiveNextInterceptorMessages:(NSUInteger)numberOfMessages {
-  if (_nextInterceptor != nil) {
-    id<GRPCInterceptorInterface> copiedNextInterceptor = _nextInterceptor;
-    dispatch_async(copiedNextInterceptor.requestDispatchQueue, ^{
-      [copiedNextInterceptor receiveNextMessages:numberOfMessages];
-    });
+  if (_nextInterceptor == nil && !_shutDown) {
+    [self createNextInterceptor];
+  }
+  if (_nextInterceptor == nil) {
+    return;
   }
+  id<GRPCInterceptorInterface> copiedNextInterceptor = _nextInterceptor;
+  dispatch_async(copiedNextInterceptor.dispatchQueue, ^{
+    [copiedNextInterceptor receiveNextMessages:numberOfMessages];
+  });
 }
 
 // Methods to forward GRPCResponseHandler callbacks to the previous object
 
 /** Forward initial metadata to the previous interceptor in the chain */
-- (void)forwardPreviousInterceptorWithInitialMetadata:(nullable NSDictionary *)initialMetadata {
-  if ([_previousInterceptor respondsToSelector:@selector(didReceiveInitialMetadata:)]) {
-    id<GRPCResponseHandler> copiedPreviousInterceptor = _previousInterceptor;
-    dispatch_async(copiedPreviousInterceptor.dispatchQueue, ^{
-      [copiedPreviousInterceptor didReceiveInitialMetadata:initialMetadata];
-    });
+- (void)forwardPreviousInterceptorWithInitialMetadata:(NSDictionary *)initialMetadata {
+  if (_nextInterceptor == nil && !_shutDown) {
+    [self createNextInterceptor];
   }
+  if (_nextInterceptor == nil) {
+    return;
+  }
+  id<GRPCResponseHandler> copiedPreviousInterceptor = _previousInterceptor;
+  dispatch_async(copiedPreviousInterceptor.dispatchQueue, ^{
+    [copiedPreviousInterceptor didReceiveInitialMetadata:initialMetadata];
+  });
 }
 
 /** Forward a received message to the previous interceptor in the chain */
 - (void)forwardPreviousInterceptorWithData:(id)data {
-  if ([_previousInterceptor respondsToSelector:@selector(didReceiveData:)]) {
-    id<GRPCResponseHandler> copiedPreviousInterceptor = _previousInterceptor;
-    dispatch_async(copiedPreviousInterceptor.dispatchQueue, ^{
-      [copiedPreviousInterceptor didReceiveData:data];
-    });
+  if (_previousInterceptor == nil) {
```

Comment:
Here an in other places in this class, can we assert that `_previousInterceptor != nil`, or atleast log an error? If we just return then we're silently dropping messages.

Reply:
It could happen that a message is forwarded after `_previousInterceptor` becomes nil. For example, if an interceptor gets a cancel request, while at the same time gets a response data, and the cancel won the race, the response should be dropped silently without an error message.